TypeSystem/Serialization/test.cc

```diff
@@ -540,6 +540,49 @@ TEST(Serialization, JSONForCppTypes) {
   }
 }
 
+TEST(Serialization, OptionalAsBinary) {
+  using namespace serialization_test;
+
+  const std::string tmp_file = bricks::FileSystem::GenTmpFileName();
+  const auto tmp_file_remover = bricks::FileSystem::ScopedRmFile(tmp_file);
+  {
+    std::ofstream ofs(tmp_file);
+    WithOptional with_optional;
+    SaveIntoBinary(ofs, with_optional);
+
+    with_optional.i = 42;
+    SaveIntoBinary(ofs, with_optional);
+
+    with_optional.b = true;
+    SaveIntoBinary(ofs, with_optional);
+
+    with_optional.i = nullptr;
+    SaveIntoBinary(ofs, with_optional);
+  }
+  {
+    std::ifstream ifs(tmp_file);
+    const auto parsed_empty = LoadFromBinary<WithOptional>(ifs);
```

Comment:
Nit: Wrapping per-deserialized tests into individual subscopes would eliminate the room for mistakenly comparing the result of the 2nd read to the result of the 1st one.